File: DevExpress-Demo/CS/NavigationDemo.Wpf/ViewModels/FileSearchViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel.DataAnnotations;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;

namespace NavigationDemo {
    public class FileSearchViewModel {
        #region Static

        static IEnumerable<DirectoryInfo> GetNestedDirectories(IEnumerable<DirectoryInfo> directories, Func<DirectoryInfo, bool> predicate) {
            return directories.SelectMany(x => {
                var childDirectories = new DirectoryInfo[0];
                try {
                    childDirectories = x.GetDirectories().Where(predicate).ToArray();
                } 
                catch { }
                return new[] { x }.Concat(GetNestedDirectories(childDirectories, predicate));
            });
        }

        #endregion

        const int maximumFilesCount = 2000;
        readonly ObservableCollection<string> _selectedPaths;
        ModifiedIntervalKind dateTimeKind = ModifiedIntervalKind.Undefined;
        #region Properties

        protected virtual IFolderBrowserDialogService FolderBrowserDialogService { get { return this.GetService<IFolderBrowserDialogService>(); } }

        public virtual string SearchPattern { get; set; }

        public ObservableCollection<string> SelectedPaths { get { return _selectedPaths; } }
        public virtual string SelectedPath { get; set; }

        public virtual SpecifiedDateKind DateKind { get; set; }

        public virtual DateTime FromDate { get; set; }
        public virtual DateTime ToDate { get; set; }

        public virtual bool IsSearchSubfolders { get; set; }
        public virtual bool IsSearchHiddenFilesAndFolders { get; set; }
        public virtual bool IsSearchSystemFolders { get; set; }
        public virtual bool IsReadOnlyFiles { get; set; }

        public virtual bool Searching { get; set; }
        [DevExpress.Mvvm.DataAnnotations.BindableProperty]
        public virtual ObservableCollection<FileInfo> SearchResult { get; set; }

        #endregion

        public FileSearchViewModel() {
            _selectedPaths = new ObservableCollection<string>();
            FromDate = DateTime.Now;
            ToDate = DateTime.Now;
            IsSearchSystemFolders = true;
            IsSearchSubfolders = true;
        }

        #region Command Methods

        public void OnInitialized() {
            SearchPattern = ".exe";
            SetSelectedPath(Directory.GetParent(Directory.GetCurrentDirectory()).FullName);
            this.GetAsyncCommand(x => x.StartSearch()).Execute(null);
        }
        public void OnUnloaded() {
            if(CanCancel())
                Cancel();
            SearchResult.Clear();
            _selectedPaths.Clear();
            SelectedPath = null;
        }
        public virtual void SelectPath() {
            if(FolderBrowserDialogService.ShowDialog()) {
                SetSelectedPath(FolderBrowserDialogService.ResultPath);
            }
        }
        public void SetSelectedPath(string path) {
            SelectedPaths.Add(path);
            SelectedPath = path;
        }
        public Task StartSearch() {
            Searching = true;
            return Task.Factory.StartNew(() => {
                WalkFolderTree();
                Searching = false;
            });
        }
        public virtual bool CanStartSearch() {
            return !(string.IsNullOrEmpty(SearchPattern) || string.IsNullOrEmpty(SelectedPath) || Searching);
        }
        public virtual void Cancel() {
            this.GetAsyncCommand(x => x.StartSearch()).CancelCommand.Execute(null);
        }
        public virtual bool CanCancel() {
            return Searching;
        }
        public virtual void SelectDateTimeKind(ModifiedIntervalKind kind) {
            dateTimeKind = kind;
        }

        #endregion
        #region Filtering Predicates

        bool AllowFolderIterate(FileSystemInfo info) {
            return GetHiddenElementPredicate(info) && (IsSearchSystemFolders || !info.Attributes.HasFlag(FileAttributes.System));
        }
        bool GetHiddenElementPredicate(FileSystemInfo info) {
            return IsSearchHiddenFilesAndFolders || !info.Attributes.HasFlag(FileAttributes.Hidden);
        }

        bool GetDateTimePredicate(FileInfo info) {
            var result = true;
            DateTime? start = null;
            DateTime? finish = null;
            GetDateTimeBounds(out start, out finish);
            if(start.HasValue && finish.HasValue)
                if(dateTimeKind != ModifiedIntervalKind.SpecifiedDates || DateKind == SpecifiedDateKind.Modified)
                    result = DateTimeIsInRange(info.LastWriteTime, start, finish);
                else
                    result = DateTimeIsInRange(DateKind == SpecifiedDateKind.Accessed ? info.LastAccessTime : info.CreationTime, start, finish);
            return result;
        }
        void GetDateTimeBounds(out DateTime? start, out DateTime? finish) {
            start = null;
            finish = null;
            switch(dateTimeKind) {
                case ModifiedIntervalKind.LastWeek:
                    start = DateTime.Now - TimeSpan.FromDays(7);
                    finish = DateTime.Now;
                    break;
                case ModifiedIntervalKind.PastMonth:
                    int year = DateTime.Now.Month > 1 ? DateTime.Now.Year : DateTime.Now.Year - 1;
                    int month = DateTime.Now.Month > 1 ? DateTime.Now.Month - 1 : 12;
                    start = new DateTime(year, month, 1);
                    finish = new DateTime(year, month, DateTime.DaysInMonth(year, month));
                    break;
                case ModifiedIntervalKind.PastYear:
                    start = DateTime.Now - TimeSpan.FromDays(365);
                    finish = DateTime.Now;
                    break;
                case ModifiedIntervalKind.SpecifiedDates:
                    start = FromDate;
                    finish = ToDate;
                    break;
                default:
                    break;
            }
        }
        bool DateTimeIsInRange(DateTime dateTime, DateTime? start, DateTime? finish) {
            return dateTime >= start.Value && dateTime <= finish.Value;
        }

        #endregion

        void WalkFolderTree() {
            var directories = GetSearchDirectories();
            var searchResult = new List<FileInfo>();
            foreach(var directoryInfo in directories) {
                if(this.GetAsyncCommand(y => y.StartSearch()).IsCancellationRequested) break;
                if(!AllowFolderIterate(directoryInfo)) continue;

                IEnumerable<FileInfo> files = null;
                try {
                    files = directoryInfo.GetFiles(string.Format("*{0}*", SearchPattern), SearchOption.TopDirectoryOnly)
                            .Where(x => GetDateTimePredicate(x) && GetHiddenElementPredicate(x) && (!IsReadOnlyFiles || x.IsReadOnly)).ToArray();
                }
                catch {
                    continue;
                }

                searchResult.AddRange(files);
                if(searchResult.Count > maximumFilesCount)
                    Cancel();
            }
            SearchResult = new ObservableCollection<FileInfo>(searchResult);
        }
        IEnumerable<DirectoryInfo> GetSearchDirectories() {
            var rootInfo = new[] { new DirectoryInfo(SelectedPath) };
            if(!IsSearchSubfolders)
                return rootInfo;
            return GetNestedDirectories(rootInfo, x => GetHiddenElementPredicate(x));
        }
    }

    public enum SpecifiedDateKind {
        [Display(Name = "Modified Date")]
        Modified,
        [Display(Name = "Created Date")]
        Created,
        [Display(Name = "Accessed Date")]
        Accessed
    }

    public enum ModifiedIntervalKind {
        Undefined,
        LastWeek,
        PastMonth,
        PastYear,
        SpecifiedDates
    }
}

File: DevExpress-Demo/VB/NavigationDemo.Wpf/ViewModels/FileSearchViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel.DataAnnotations
Imports System.IO
Imports System.Linq
Imports System.Threading.Tasks
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports System.Runtime.InteropServices

Namespace NavigationDemo

    Public Class FileSearchViewModel

#Region "Static"
        Private Shared Function GetNestedDirectories(ByVal directories As IEnumerable(Of DirectoryInfo), ByVal predicate As Func(Of DirectoryInfo, Boolean)) As IEnumerable(Of DirectoryInfo)
            Return directories.SelectMany(Function(x)
                Dim childDirectories = New DirectoryInfo(-1) {}
                Try
                    childDirectories = x.GetDirectories().Where(predicate).ToArray()
                Catch
                End Try

                Return {x}.Concat(GetNestedDirectories(childDirectories, predicate))
            End Function)
        End Function

#End Region
        Const maximumFilesCount As Integer = 2000

        Private ReadOnly _selectedPaths As ObservableCollection(Of String)

        Private dateTimeKind As ModifiedIntervalKind = ModifiedIntervalKind.Undefined

#Region "Properties"
        Protected Overridable ReadOnly Property FolderBrowserDialogService As IFolderBrowserDialogService
            Get
                Return GetService(Of IFolderBrowserDialogService)()
            End Get
        End Property

        Public Overridable Property SearchPattern As String

        Public ReadOnly Property SelectedPaths As ObservableCollection(Of String)
            Get
                Return _selectedPaths
            End Get
        End Property

        Public Overridable Property SelectedPath As String

        Public Overridable Property DateKind As SpecifiedDateKind

        Public Overridable Property FromDate As Date

        Public Overridable Property ToDate As Date

        Public Overridable Property IsSearchSubfolders As Boolean

        Public Overridable Property IsSearchHiddenFilesAndFolders As Boolean

        Public Overridable Property IsSearchSystemFolders As Boolean

        Public Overridable Property IsReadOnlyFiles As Boolean

        Public Overridable Property Searching As Boolean

        <DataAnnotations.BindableProperty>
        Public Overridable Property SearchResult As ObservableCollection(Of FileInfo)

#End Region
        Public Sub New()
            _selectedPaths = New ObservableCollection(Of String)()
            FromDate = Date.Now
            ToDate = Date.Now
            IsSearchSystemFolders = True
            IsSearchSubfolders = True
        End Sub

#Region "Command Methods"
        Public Sub OnInitialized()
            SearchPattern = ".exe"
            SetSelectedPath(Directory.GetParent(Directory.GetCurrentDirectory()).FullName)
            POCOViewModelExtensions.GetAsyncCommand(Me, Function(x) x.StartSearch()).Execute(Nothing)
        End Sub

        Public Sub OnUnloaded()
            If CanCancel() Then Cancel()
            SearchResult.Clear()
            _selectedPaths.Clear()
            SelectedPath = Nothing
        End Sub

        Public Overridable Sub SelectPath()
            If FolderBrowserDialogService.ShowDialog() Then
                SetSelectedPath(FolderBrowserDialogService.ResultPath)
            End If
        End Sub

        Public Sub SetSelectedPath(ByVal path As String)
            SelectedPaths.Add(path)
            SelectedPath = path
        End Sub

        Public Function StartSearch() As Task
            Searching = True
            Return Task.Factory.StartNew(Sub()
                WalkFolderTree()
                Searching = False
            End Sub)
        End Function

        Public Overridable Function CanStartSearch() As Boolean
            Return Not(String.IsNullOrEmpty(SearchPattern) OrElse String.IsNullOrEmpty(SelectedPath) OrElse Searching)
        End Function

        Public Overridable Sub Cancel()
            POCOViewModelExtensions.GetAsyncCommand(Me, Function(x) x.StartSearch()).CancelCommand.Execute(Nothing)
        End Sub

        Public Overridable Function CanCancel() As Boolean
            Return Searching
        End Function

        Public Overridable Sub SelectDateTimeKind(ByVal kind As ModifiedIntervalKind)
            dateTimeKind = kind
        End Sub

#End Region
#Region "Filtering Predicates"
        Private Function AllowFolderIterate(ByVal info As FileSystemInfo) As Boolean
            Return GetHiddenElementPredicate(info) AndAlso (IsSearchSystemFolders OrElse Not info.Attributes.HasFlag(FileAttributes.System))
        End Function

        Private Function GetHiddenElementPredicate(ByVal info As FileSystemInfo) As Boolean
            Return IsSearchHiddenFilesAndFolders OrElse Not info.Attributes.HasFlag(FileAttributes.Hidden)
        End Function

        Private Function GetDateTimePredicate(ByVal info As FileInfo) As Boolean
            Dim result = True
            Dim start As Date? = Nothing
            Dim finish As Date? = Nothing
            GetDateTimeBounds(start, finish)
            If start.HasValue AndAlso finish.HasValue Then
                If dateTimeKind <> ModifiedIntervalKind.SpecifiedDates OrElse DateKind = SpecifiedDateKind.Modified Then
                    result = DateTimeIsInRange(info.LastWriteTime, start, finish)
                Else
                    result = DateTimeIsInRange(If(DateKind = SpecifiedDateKind.Accessed, info.LastAccessTime, info.CreationTime), start, finish)
                End If
            End If

            Return result
        End Function

        Private Sub GetDateTimeBounds(<Out> ByRef start As Date?, <Out> ByRef finish As Date?)
            start = Nothing
            finish = Nothing
            Select Case dateTimeKind
                Case ModifiedIntervalKind.LastWeek
                    start = Date.Now - TimeSpan.FromDays(7)
                    finish = Date.Now
                Case ModifiedIntervalKind.PastMonth
                    Dim year As Integer = If(Date.Now.Month > 1, Date.Now.Year, Date.Now.Year - 1)
                    Dim month As Integer = If(Date.Now.Month > 1, Date.Now.Month - 1, 12)
                    start = New DateTime(year, month, 1)
                    finish = New DateTime(year, month, Date.DaysInMonth(year, month))
                Case ModifiedIntervalKind.PastYear
                    start = Date.Now - TimeSpan.FromDays(365)
                    finish = Date.Now
                Case ModifiedIntervalKind.SpecifiedDates
                    start = FromDate
                    finish = ToDate
                Case Else
            End Select
        End Sub

        Private Function DateTimeIsInRange(ByVal dateTime As Date, ByVal start As Date?, ByVal finish As Date?) As Boolean
            Return dateTime >= start.Value AndAlso dateTime <= finish.Value
        End Function

#End Region
        Private Sub WalkFolderTree()
            Dim directories = GetSearchDirectories()
            Dim searchResult = New List(Of FileInfo)()
            For Each directoryInfo In directories
                If POCOViewModelExtensions.GetAsyncCommand(Me, Function(y) y.StartSearch()).IsCancellationRequested Then Exit For
                If Not AllowFolderIterate(directoryInfo) Then Continue For
                Dim files As IEnumerable(Of FileInfo) = Nothing
                Try
                    files = directoryInfo.GetFiles(String.Format("*{0}*", SearchPattern), SearchOption.TopDirectoryOnly).Where(Function(x) GetDateTimePredicate(x) AndAlso GetHiddenElementPredicate(x) AndAlso (Not IsReadOnlyFiles OrElse x.IsReadOnly)).ToArray()
                Catch
                    Continue For
                End Try

                searchResult.AddRange(files)
                If searchResult.Count > maximumFilesCount Then Cancel()
            Next

            Me.SearchResult = New ObservableCollection(Of FileInfo)(searchResult)
        End Sub

        Private Function GetSearchDirectories() As IEnumerable(Of DirectoryInfo)
            Dim rootInfo = {New DirectoryInfo(SelectedPath)}
            If Not IsSearchSubfolders Then Return rootInfo
            Return GetNestedDirectories(rootInfo, Function(x) GetHiddenElementPredicate(x))
        End Function
    End Class

    Public Enum SpecifiedDateKind
        <Display(Name:="Modified Date")>
        Modified
        <Display(Name:="Created Date")>
        Created
        <Display(Name:="Accessed Date")>
        Accessed
    End Enum

    Public Enum ModifiedIntervalKind
        Undefined
        LastWeek
        PastMonth
        PastYear
        SpecifiedDates
    End Enum
End Namespace
